File: spring-webflux-plain-java-example/src/test/java/de/cfranzen/examples/webflux/groups/CustomerGroupClientTest.java
package de.cfranzen.examples.webflux.groups;

import com.github.tomakehurst.wiremock.common.Json;
import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo;
import com.github.tomakehurst.wiremock.junit5.WireMockTest;
import org.junit.jupiter.api.Test;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.List;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static org.assertj.core.api.Assertions.assertThat;

@WireMockTest
class CustomerGroupClientTest {

    @Test
    void testRetrievingCustomerGroup(final WireMockRuntimeInfo wmRuntimeInfo) {
        // Given
        final var customerGroup = new CustomerGroup(
                "testGroup",
                List.of(
                        new CustomerGroupEntry("Peter", "Mueller"),
                        new CustomerGroupEntry("Sabine", "Hochhaus"),
                        new CustomerGroupEntry("Heike", "Recker")
                )
        );
        stubCustomerGroupServer(customerGroup);

        // When
        final var result = buildClient(wmRuntimeInfo).getCustomerGroup(customerGroup.name());

        // Then
        assertThat(result.block())
                .usingRecursiveComparison()
                .isEqualTo(customerGroup);

    }

    private CustomerGroupClient buildClient(final WireMockRuntimeInfo wmRuntimeInfo) {
        final var webClient = WebClient.builder().baseUrl(wmRuntimeInfo.getHttpBaseUrl()).build();
        return new CustomerGroupClient(webClient);
    }

    private void stubCustomerGroupServer(final CustomerGroup result) {
        stubFor(get("/api/customerGroup/" + result.name())
                .willReturn(okJson(Json.write(result))));
    }
}

File: spring-webflux-kotlin-example/src/test/kotlin/de/cfranzen/examples/webflux/kotlin/groups/CustomerGroupClientTest.kt
package de.cfranzen.examples.webflux.kotlin.groups

import com.github.tomakehurst.wiremock.client.WireMock.get
import com.github.tomakehurst.wiremock.client.WireMock.okJson
import com.github.tomakehurst.wiremock.client.WireMock.stubFor
import com.github.tomakehurst.wiremock.common.Json
import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo
import com.github.tomakehurst.wiremock.junit5.WireMockTest
import kotlinx.coroutines.test.runTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.web.reactive.function.client.WebClient

@WireMockTest
internal class CustomerGroupClientTest {

    @Test
    fun testRetrievingCustomerGroup(wmRuntimeInfo: WireMockRuntimeInfo) = runTest {
        // Given
        val customerGroup = CustomerGroup(
            name = "testGroup",
            entries = listOf(
                CustomerGroupEntry("Peter", "Mueller"),
                CustomerGroupEntry("Sabine", "Hochhaus"),
                CustomerGroupEntry("Heike", "Recker")
            )
        )
        stubCustomerGroupServer(customerGroup)

        // When
        val result = buildClient(wmRuntimeInfo).getCustomerGroup(customerGroup.name)

        // Then
        assertThat(result)
            .usingRecursiveComparison()
            .isEqualTo(customerGroup)
    }

    private fun buildClient(wmRuntimeInfo: WireMockRuntimeInfo): CustomerGroupClient {
        val webClient = WebClient.builder().baseUrl(wmRuntimeInfo.httpBaseUrl).build()
        return CustomerGroupClient(webClient)
    }

    private fun stubCustomerGroupServer(result: CustomerGroup) {
        stubFor(
            get("/api/customerGroup/" + result.name)
                .willReturn(okJson(Json.write(result)))
        )
    }
}